lib/std/fmt/scan.zig

```diff
@@ -0,0 +1,239 @@
+    return c == '0' or c == '1';
+}
+
+/// similar to scanf in c except conversion specifiers are given like "{d}" rather than "%d"
+/// valid specifiers include
+/// - {c}     - match exactly one character, don't parse. '0' -> '0' (48)
+/// - {d}, {} - match and parse a base 10 number
+/// - {e}     - match and parse a floating point number
+/// - {x}     - match and parse a hex number with optional leading "0x" or "0X"
+/// - {o}     - match and parse an octal number with optional leading "0o"
+/// - {b}     - match and parse a binary number with optional leading "0b"
+/// - {s}     - match any input until a whitespace or the following character in fmt string
+/// Notes: curly braces can be escaped in fmt "{{}}"
+///        each whitespace in fmt will skip zero or more input whitespaces
+pub fn scan(input: []const u8, comptime fmt: []const u8, args: anytype) Error!usize {
+    comptime var arg_i: usize = 0;
+    var i: usize = 0; // input index
+    const State = enum { start, specifier, specifier_end };
+    comptime var state: State = .start;
+    inline for (fmt) |fmt_c, fmt_i| {
+        var neg = false;
+        if (i == std.math.maxInt(usize)) return error.InputMatchFailure;
+        switch (state) {
+            .start => switch (fmt_c) {
+                '{' => state = .specifier,
+                // match }}
+                '}' => if (fmt_i > 0 and fmt[fmt_i - 1] == '}') {
+                    if (input[i] == '}') {
+                        i += 1;
+                    } else i = std.math.maxInt(usize);
+                },
+                else => if (std.ascii.isSpace(fmt_c)) {
+                    while (i < input.len and std.ascii.isSpace(input[i])) : (i += 1) {}
+                } else if (i < input.len and fmt_c == input[i]) {
+                    i += 1;
+                } else {
+                    i = std.math.maxInt(usize);
+                },
+            },
+            .specifier_end => switch (fmt_c) {
+                '}' => state = .start,
+                else => return error.UnknownSpecifier,
+            },
+            .specifier => {
+                // special case escaped {{
+                if (fmt_c == '{') {
+                    if (i < input.len and fmt_c == input[i]) {
+                        i += 1;
+                        state = .start;
+                    } else i = std.math.maxInt(usize);
+                    continue;
+                }
+
+                if (i < input.len) {
+                    if (input[i] == '-') {
+                        neg = true;
+                        i += 1;
+                    } else if (input[i] == '+') {
+                        i += 1;
+                    }
+                }
+                if (arg_i >= args.len) return error.TooFewArguments;
+                var arg = args[arg_i];
+                switch (fmt_c) {
+                    '}', 'd' => {
+                        if (i < input.len) i += try parseInt(input[i..], 10, std.ascii.isDigit, arg, neg);
```

Comment:
You should split the scanning from the parsing, the latter is done by `fmt.parseInt` so you only have to do the former here.

Write some small helper routines such as `scanSigned`, `scanUnsigned`, `scanFloat` (remember of "weird" values such as `NaN`, `infinite` and their different casing/abbreviations), `scanBool` to return a slice of text to parse and update the current input position.

Reply:
This still needs work as i'm still mixing the scanning and parsing in new commit (look for is_negative).